mllib/src/main/scala/org/apache/spark/ml/classification/NaiveBayes.scala

```diff
@@ -0,0 +1,192 @@
+   * Default is "multinomial"
+   */
+  def setModelType(value: String): this.type = set(modelType, value)
+  setDefault(modelType -> "multinomial")
+
+  override protected def train(dataset: DataFrame): NaiveBayesModel = {
+    val oldDataset: RDD[LabeledPoint] = extractLabeledPoints(dataset)
+    val oldModel = OldNaiveBayes.train(oldDataset, $(lambda), $(modelType))
+    NaiveBayesModel.fromOld(oldModel, this)
+  }
+
+  override def copy(extra: ParamMap): NaiveBayes = defaultCopy(extra)
+}
+
+/**
+ * Model produced by [[NaiveBayes]]
+ */
+class NaiveBayesModel private[ml] (
+    override val uid: String,
+    val labels: Vector,
+    val pi: Vector,
+    val theta: Matrix,
+    val modelType: String)
+  extends PredictionModel[Vector, NaiveBayesModel] {
+
+  import NaiveBayesModel.{Bernoulli, Multinomial, supportedModelTypes}
+
+  require(supportedModelTypes.contains(modelType),
+    s"NaiveBayes was created with an unknown modelType: $modelType.")
+
+  /**
+   * Bernoulli scoring requires log(condprob) if 1, log(1-condprob) if 0.
+   * This precomputes log(1.0 - exp(theta)) and its sum which are used for the linear algebra
+   * application of this condition (in predict function).
+   */
+  private val (thetaMinusNegTheta, negThetaSum) = modelType match {
+    case Multinomial => (None, None)
+    case Bernoulli =>
+      val negTheta = theta.map(value => math.log(1.0 - math.exp(value)))
+      val ones = new DenseVector(Array.fill(theta.numCols){1.0})
+      val thetaMinusNegTheta = theta.map { value =>
+        value - math.log(1.0 - math.exp(value))
+      }
+      (Option(thetaMinusNegTheta), Option(negTheta.multiply(ones)))
+    case _ =>
+      // This should never happen.
+      throw new UnknownError(s"Invalid modelType: $modelType.")
+  }
+
+  override protected def predict(features: Vector): Double = {
+    modelType match {
+      case Multinomial =>
+        val prob = theta.multiply(features)
+        BLAS.axpy(1.0, pi, prob)
+        labels(prob.argmax)
+      case Bernoulli =>
+        features.foreachActive{ (index, value) =>
+          if (value != 0.0 && value != 1.0) {
+            throw new SparkException(
+              s"Bernoulli naive Bayes requires 0 or 1 feature values but found $features")
+          }
+        }
+        val prob = thetaMinusNegTheta.get.multiply(features)
+        BLAS.axpy(1.0, pi, prob)
+        BLAS.axpy(1.0, negThetaSum.get, prob)
+        labels(prob.argmax)
+      case _ =>
+        // This should never happen.
+        throw new UnknownError(s"Invalid modelType: $modelType.")
+    }
+  }
+
+  override def copy(extra: ParamMap): NaiveBayesModel = {
+    copyValues(new NaiveBayesModel(uid, labels, pi, theta, modelType), extra)
+  }
+
+  override def toString: String = {
+    s"NaiveBayesModel with ${labels.size} classes"
+  }
+
+}
+
+private[ml] object NaiveBayesModel {
+
+  /** String name for multinomial model type. */
+  private[classification] val Multinomial: String = "multinomial"
```

Comment:
Could you please follow the spark.mllib example:
* Put these under NaiveBayes (not Model)
* Refer to the old NaiveBayes values, rather than re-writing them here (just to make sure we use exactly the same names).